File: p976e/largest_perimeter.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int largestPerimeter(vector<int>& nums)
    {
        sort(nums.begin(), nums.end(), [](int a, int b) -> bool { return a > b; });
        for (int i = 2; i < nums.size(); ++i) {
            if (nums[i] + nums[i - 1] > nums[i - 2])
                return nums[i] + nums[i - 1] + nums[i - 2];
        }
        return 0;
    }
};
// TESTS
struct Test {
    vector<int> input;
    int expected;
    void run()
    {
        Solution sol;
        int actual = sol.largestPerimeter(input);
        cout << "Largest perimeter of a triangle -> " << actual << endl;
        assert(actual == expected);
    }
};
int main(int argc, char* argv[])
{
    vector<Test> tests = {
        { { 2, 1, 2 }, 5 },
        { { 1, 2, 1 }, 0 },
        { { 3, 2, 3, 4 }, 10 },
        { { 3, 6, 2, 3 }, 8 }
    };
    for (auto& t : tests)
        t.run();
    return 0;
}

File: p976e/largest_perimeter.py
class Solution:
    def largestPerimeter(self, nums: list[int]) -> int:
        ns = sorted(nums, reverse=True)
        for i in range(2, len(nums)):
            if ns[i] + ns[i - 1] > ns[i - 2]:
                return sum(ns[i - 2 : i + 1])
        return 0


# TESTS
for nums, expected in [
    ([2, 1, 2], 5),
    ([1, 2, 1], 0),
    ([3, 2, 3, 4], 10),
    ([3, 6, 2, 3], 8),
]:
    sol = Solution()
    actual = sol.largestPerimeter(nums)
    print("Largest perimeter in", nums, "->", actual)
    assert actual == expected
